### ISSUE TITLE
Make it use docker-api.el

### ISSUE THREAD

**Issue by Silex** [Silex/docker.el#2]
This is harder than it looks, because of several issues:

- images: it requires parsing them differently:
  - "docker images" returns one line per "repo:tag" combination, even if the image id is the same
  - "GET images/json" returns one hash per image, and has a "repo_tags" key which contains all its aliases
- run: people expect the popup to show the classic "run" options (-it, --rm, etc). Offering these options but then having to translate them using the api increases the risk of having subtle different behaviors than from the command line experience.

Basically, an implementation using the API is not straighforward, and is likely to differ from the console experience. Magit's success is also because it stays close to the git console experience (e.g it calls "git foo" constantly).

**Comment by Silex** [comment]
@Yuki-Inoue: if you want, you can test the `docker-api` branch. It still needs some polish, but the containers listing uses the docker-api now. You need to load some lisp files from https://github.com/Silex/docker-api.el for it to work.

What needs to still be done is parse the date correctly, etc

**Comment by Yuki-Inoue** [comment]
@Silex Ah,,, I'm starting to figure it out.  So the docker supports native json API.  I'll try your `docker-api.el` when I have some time.  :)

**Comment by Silex** [comment]
@Yuki-Inoue: ok... I'll probably stop my effort in using docker-api, because today I found `docker ps --format "table {{.ID}}\t{{.Image}}"`

That allows me to specify the format, which will greatly simplify the parsing of the various docker commands. Also because implementing "docker run" with the docker-api is very complicated, and I don't feel very comfortable using the docker-api for everything EXCEPT `docker run`.

**Comment by Silex** [comment]
Until `docker-run` is doable using the docker-api, I think we won't use it.

See https://docs.docker.com/engine/reference/api/docker_remote_api_v1.23/#3-1-inside-docker-run for more details.

**Comment by Yuki-Inoue** [comment]
@Silex I get it, so you are saying we can specify format for `docker ps` and can use '\t' as the separator.  Sounds good to me, given we would parse the `command` as the last column, which I believe to be the only attribute which might contain '\t' character.

**Comment by Silex** [comment]
@Yuki-Inoue: good catch about "command" that may contain a `\t`.

I'll add a `TODO` for now but if you want to fix it feel free.

**Comment by tzz** [comment]
@Yuki-Inoue looks like other people had the same issues: http://blog.bordage.pro/avoid-docker-py/

For the output formatting, note you can encapsulate things in JSON inside the Go template: `docker ps --format '[{{.ID|json}},{{.Image|json}},{{.Command|json}}]'` would give you a parseable JSON array.

**Comment by Silex** [comment]
@tzz: oh, cute. I might switch to that indeed

**Comment by Yuki-Inoue** [comment]
@tzz 👍
